src/snmalloc/mem/secondary/gwp_asan.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "gwp_asan/guarded_pool_allocator.h"
+#include "snmalloc/ds_core/defines.h"
+#include "snmalloc/mem/sizeclasstable.h"
+#if defined(SNMALLOC_BACKTRACE_HEADER)
+#  include SNMALLOC_BACKTRACE_HEADER
+#endif
+
+namespace snmalloc
+{
+  class GwpAsanSecondaryAllocator
+  {
+    static inline gwp_asan::GuardedPoolAllocator singleton;
+    static inline size_t max_allocation_size;
+
+  public:
+    static void initialize() noexcept
+    {
+      // for now, we use default options
+      gwp_asan::options::Options opt;
+      opt.setDefaults();
+#ifdef SNMALLOC_BACKTRACE_HEADER
+      opt.Backtrace = [](uintptr_t* buf, size_t length) {
+        return static_cast<size_t>(
+          ::backtrace(reinterpret_cast<void**>(buf), static_cast<int>(length)));
+      };
+#endif
+      singleton.init(opt);
+      max_allocation_size =
+        singleton.getAllocatorState()->maximumAllocationSize();
+    }
+
+    SNMALLOC_FAST_PATH static void* allocate(size_t size)
+    {
+      if (SNMALLOC_UNLIKELY(singleton.shouldSample()))
```

Comment:
As this is only getting called when a free list is exhausted should we be doing something else here.  I.e. should we be passing how many requests of this size have been handled since we last called this allocator?

Reply:
`shouldSample` is implemented as
```
    // NextSampleCounter == 0 means we "should regenerate the counter".
    //                   == 1 means we "should sample this allocation".
    // AdjustedSampleRatePlusOne is designed to intentionally underflow. This
    // class must be valid when zero-initialised, and we wish to sample as
    // infrequently as possible when this is the case, hence we underflow to
    // UINT32_MAX.
    if (GWP_ASAN_UNLIKELY(getThreadLocals()->NextSampleCounter == 0))
      getThreadLocals()->NextSampleCounter =
          ((getRandomUnsigned32() % (AdjustedSampleRatePlusOne - 1)) + 1) &
          ThreadLocalPackedVariables::NextSampleCounterMask;

    return GWP_ASAN_UNLIKELY(--getThreadLocals()->NextSampleCounter == 0);
```

Reply:
Sorry, I meant.  Normally usage of GWP Asan, every call to `malloc` will call `shouldSample`.  But this usage has moved it from the fast path, so we will call it every new free list for that size.  This in the case of 16 byte allocations worst case could be 1024 allocations between calls.  This will alter the likelihood of it being called, and might bias away from small allocations.  I think for GWP Asan on we would want randomisation too, so that the free list that gets exhausted.

I think this is probably the same information that would be required to unbias profiling, so should probably be left to a subsequent PR, but we should perhaps add either an issue or a comment about this.